src/engine/QueryPlanner.cpp

```diff
@@ -2029,18 +2042,63 @@ auto QueryPlanner::createSpatialJoin(
   return plan;
 }
 
+// _____________________________________________________________________________________________________________________
+auto QueryPlanner::createJoinWithUnitedTransitivePath(
+    SubtreePlan a, SubtreePlan b,
+    const std::vector<std::array<ColumnIndex, 2>>& jcs) const
+    -> std::vector<SubtreePlan> {
+  auto aUnionPath =
+      std::dynamic_pointer_cast<Union>(a._qet->getRootOperation());
+  auto bUnionPath =
+      std::dynamic_pointer_cast<Union>(b._qet->getRootOperation());
+  std::vector<SubtreePlan> candidates{};
+  auto findCandidates = [this, &candidates, &jcs](Union& unionOperation,
+                                                  const SubtreePlan& other,
+                                                  bool unionIsLeft) {
+    auto* qec = other._qet->getRootOperation()->getExecutionContext();
+
+    // Clone `other` once, to avoid using it twice in the same query plan.
+    SubtreePlan clonedOther = other;
+    clonedOther._qet = other._qet->clone();
+    SubtreePlan tmpPlan = other;
+    const auto& leftCandidate = unionIsLeft ? tmpPlan : other;
+    const auto& rightCandidate = unionIsLeft ? clonedOther : tmpPlan;
+
+    tmpPlan._qet = unionOperation.leftChild();
+    auto joinedLeft = createJoinCandidates(leftCandidate, rightCandidate, jcs);
+    tmpPlan._qet = unionOperation.rightChild();
+    auto joinedRight = createJoinCandidates(leftCandidate, rightCandidate, jcs);
+
+    for (const auto& leftPlan : joinedLeft) {
+      for (const auto& rightPlan : joinedRight) {
+        SubtreePlan candidate =
+            makeSubtreePlan<Union>(qec, leftPlan._qet, rightPlan._qet);
+        mergeSubtreePlanIds(candidate, leftPlan, rightPlan);
```

Comment:
I think that is not quite correct, as the planning of `UNION`s has their own recursive subtreePlanId Space (unfortunately currently this whole subtreePlanId).
I think the correct plan Ids for all the union would be:
"the subtree plan ids of the Original union (before you crack it open) merged with the plan Id(s) of the join partner".